scipy/stats/_qmc_cy.pyx

```diff
@@ -0,0 +1,383 @@
+        return mixture_discrepancy(sample, iterative, workers)
+
+
+def _cy_wrapper_l2_star_discrepancy(double[:, ::1] sample,
+                                   bint iterative, workers):
+        return l2_star_discrepancy(sample, iterative, workers)
+
+
+cdef double centered_discrepancy(double[:, ::1] sample_view,
+                             bint iterative, unsigned int workers):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0
+        double prod, disc1 = 0
+
+    for i in range(n):
+        prod = 1
+        for j in range(d):
+                prod *= (
+                        1 + 0.5 * fabs(sample_view[i, j] - 0.5) - 0.5 * fabs(
+                        sample_view[i, j] - 0.5) ** 2
+                )
+        disc1 += prod
+
+    cdef double disc2 = threaded_loops(centered_discrepancy_loop, sample_view,
+                                       workers)
+
+    if iterative:
+        n += 1
+
+    return ((13.0 / 12.0) ** d - 2.0 / n * disc1 +
+                1.0 / (n ** 2) * disc2)
+
+
+cdef double centered_discrepancy_loop(double[:, ::1] sample_view,
+                                    Py_ssize_t istart, Py_ssize_t istop) nogil:
+
+    cdef:
+        Py_ssize_t i, j, k
+        double prod, disc2 = 0
+
+    for i in range(istart, istop):
+        for j in range(sample_view.shape[0]):
+            prod = 1
+            for k in range(sample_view.shape[1]):
+                    prod *= (
+                            1 + 0.5 * fabs(sample_view[i, k] - 0.5)
+                            + 0.5 * fabs(sample_view[j, k] - 0.5)
+                            - 0.5 * fabs(sample_view[i, k] - sample_view[j, k])
+                    )
+            disc2 += prod
+
+    return disc2
+
+
+cdef double wrap_around_discrepancy(double[:, ::1] sample_view,
+                             bint iterative, unsigned int workers):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double x_kikj, prod = 1, disc
+
+    disc = threaded_loops(wrap_around_loop, sample_view,
+                          workers)
+
+    if iterative:
+        n += 1
+
+    return - (4.0 / 3.0) ** d + 1.0 / (n ** 2) * disc
+
+
+cdef double wrap_around_loop(double[:, ::1] sample_view,
+                                    Py_ssize_t istart, Py_ssize_t istop) nogil:
+
+    cdef:
+        Py_ssize_t i, j, k
+        double prod, disc = 0
+
+    for i in range(istart, istop):
+        for j in range(sample_view.shape[0]):
+            prod = 1
+            for k in range(sample_view.shape[1]):
+                x_kikj = fabs(sample_view[i, k] - sample_view[j, k])
+                prod *= 3.0 / 2.0 - x_kikj + x_kikj ** 2
+            disc += prod
+
+    return disc
+
+
+cdef double mixture_discrepancy(double[:, ::1] sample_view,
+                             bint iterative, unsigned int workers):
```

Comment:
Is there a reason why this function and `threaded_loops` are not using `nogil`? I haven't tried, but I don't see anything needing the CPython API.

Reply:
Is there a performance gain in using `nogil`everywhere ? I will change it though, because you are right, there is no call from the Cython API.